py/mpconfig.h

```diff
@@ -62,6 +62,24 @@
 #include <mpconfigport.h>
 #endif
 
+// Sets the default set of features.
+// 0 -- disable all optional features (i.e. minimal port).
+// 1 -- only enable core features (constrained flash, e.g. STM32L072)
+// 2 -- enable most common features (small on-device flash, e.g. STM32F411)
+// 3 -- enable convenience features (larger on-device flash, e.g. STM32F405)
+// 4 -- enable all common features (external flash, e.g. ESP32, unix)
+// 5 -- enable everything (coverage)
+#ifndef MICROPY_CONFIG_ROM_LEVEL
+#define MICROPY_CONFIG_ROM_LEVEL (1)
+#endif
+
+// Helper macros for "have at least this level".
+#define _MICROPY_CONFIG_ROM_LEVEL_CORE (MICROPY_CONFIG_ROM_LEVEL >= 1)
+#define _MICROPY_CONFIG_ROM_LEVEL_COMMON (MICROPY_CONFIG_ROM_LEVEL >= 2)
+#define _MICROPY_CONFIG_ROM_LEVEL_CONVENIENCE (MICROPY_CONFIG_ROM_LEVEL >= 3)
+#define _MICROPY_CONFIG_ROM_LEVEL_ALL (MICROPY_CONFIG_ROM_LEVEL >= 4)
+#define _MICROPY_CONFIG_ROM_LEVEL_EVERYTHING (MICROPY_CONFIG_ROM_LEVEL >= 5)
```

Comment:
'all' and 'everything' are quite similar; use 'coverage' for the latter?

Wrt 'rom level' naming: since it's a set of feature levels, perhaps just name it like that (MICROPY_FEATURE_LEVEL) as well?

I can imagine that instead of the single-line description of which level has what, there'd be more elaborate comments. In that case the explanation of which one does what is going to be further away from where it gets defined. And now you have to repeat the numbers. How about moving those comments here, together with their respective definition? I.e. 

```
 // Only enable core features (constrained flash, e.g. STM32L072)
 #define _MICROPY_CONFIG_ROM_LEVEL_CORE (MICROPY_CONFIG_ROM_LEVEL >= 1)
```

Perhaps overkill, but people might want to adopt this principle and define their own levels which are somewhere in between existing ones, which can be done by leaving gaps here i.e. 10, 20, 30, ...

Reply:
> 'all' and 'everything' are quite similar; use 'coverage' for the latter?

Yep, I think you're right.

> Wrt 'rom level' naming: since it's a set of feature levels, perhaps just name it like that (MICROPY_FEATURE_LEVEL) as well?

I did want to also add RAM levels too (e.g. rp2 is ram but not flash constrained), but I think given that ROM generally maps to features, this makes sense. My reasoning for calling it ROM was to emphasize that we're "saving ROM", not "cutting features" (although I agree that's a fairly abstract distinction).

> Perhaps overkill, but people might want to adopt this principle and define their own levels which are somewhere in between existing ones, which can be done by leaving gaps here i.e. 10, 20, 30, ...

I guess so... perhaps this could be useful for board-specific overrides. I wanted to avoid having to decide "is this as 27 or 26 level feature discussions" in the future.

On that though, I think it's worth putting #defines for the levels (somewhere before mpconfigboard.h gets imported) so that the numbers themselves are not meaningful. (i.e. minimal/mpconfigport.h should set it to "MINIMAL" not "0").